### PR TITLE
CMakeLists.txt updates for modern CMake

### PR DESCRIPTION
**Pull request by nimish**
Update cmake to 3.21+ and simplify the logic

Uses the bundled FindODBC and uses dependent options

 - [x] Review
 - [x] Adjust for comments
 - [x] All CI builds and checks have passed

### CODE REVIEW

**Comment by mloskot** [comment]
This is great, thank you.

First, I'd like to hear from @jimhester @detule @lexicalunit  if they have any objections to bump the required version of CMake.

**Comment by nimish** [comment]
> This is great, thank you.
> 
> First, I'd like to hear from @jimhester @detule @lexicalunit if they have any objections to bump the required version of CMake.

Sure. I think 3.21 was conservative to get all the relevant policies. It could probably be relaxed.

**Comment by detule** [comment]
Hi there: took a quick look.

I probably read the diff wrong, but does this also bump the required language standard to c++17? 

Also the comment
```
# #######################################
# # require and enable C++20
# #######################################
```
seems misleading.

**Comment by nimish** [comment]
> Hi there: took a quick look.
> 
> I probably read the diff wrong, but does this also bump the required language standard to c++17?
> 
> Also the comment
> 
> ```
> # #######################################
> # # require and enable C++20
> # #######################################
> ```
> 
> seems misleading.

Oops, that was from a different branch that shouldn't have made it in here. Relaxed back to 14.

**Comment by mloskot** [comment]
Good catch @detule!

@nimish  Yes, we don't want to switch away from C++14 yet, especially there are important users like r-odbc which may still need C++14. If we switch, it will be C++17 though, see https://github.com/nanodbc/nanodbc/discussions/321 However, we can accept conditionally-compiled support for newer C++ standards.

**Comment by nimish** [comment]
> Good catch @detule!
> 
> @nimish Yes, we don't want to switch away from C++14 yet, especially there are important users like r-odbc which may still need C++14. If we switch, it will be C++17 though, see #321 However, we can accept conditionally-compiled support for newer C++ standards.

No problem, was testing out C++20 support in general. 

Can't quite understand the reason the CI is now failing

**Comment by mloskot** [comment]
> Can't quite understand the reason the CI is now failing

Sorry, I didn't have time to check. I will try to look at the CI results tomorrow and see if I can help

File: .github/workflows/ci-linux.yml [hunk 1]
```diff
@@ -56,7 +56,7 @@ jobs:
       - uses: actions/checkout@v3
       - name: Install
         run: |
-          sudo apt-get install -y cmake ${{ matrix.cxx }}
+          sudo apt-get install -y cmake libboost-locale-dev libboost-dev ${{ matrix.cxx }}
```
Comment (mloskot) [inline .github/workflows/ci-linux.yml:59]: Boost is only required if CMake configuration phase is specified with
https://github.com/nanodbc/nanodbc/blob/5b52ff0c8fd70572ed610bab8104fd5be08d954e/CMakeLists.txt#L19

File: example/empty.cpp [hunk 2]
```diff
@@ -1,4 +1,3 @@
-#include "example_unicode_utils.h"
```
Comment (mloskot) [inline example/empty.cpp:1]: Was the file renaming intentional here?

Comment (nimish) [reply example/empty.cpp:1]: Yes -- the cmake `configure_file` didn't interpolate anything as far as I could tell

File: .github/workflows/ci-linux.yml (hunk 1 above)
Comment (nimish) [reply .github/workflows/ci-linux.yml:59]: Yep, looks like I defaulted to using Boost.Locale if Unicode support was built in. Fixed

File: example/empty.cpp (hunk 2 above)
Comment (mloskot) [reply example/empty.cpp:1]: Yes, there is nothing to interpolate indeed.

The unobvious reason why it there is `empty.in.cpp` in the repo and not `empty.cpp` is to allow this workflow
- CMake creates `empty.cpp` 
- The `empty.cpp` is available to contributors and users as playground for experiments, to modify, build and debug as part of the nanodbc project.
- The `empty.cpp` is automatically ignored, so contributors don't have to be cautious to not to commit their experiments to the nanodbc repo or forks

Comment (mloskot) [reply example/empty.cpp]: Thanks @nimish 
If there are better ways to achieve similar, then please feel free to propose a change.

**Comment by nimish** [comment]
@mloskot 
* reverted empty.cpp changes
* fixed some cmake defaulting issues with vars
* added some inline comments to help explain some of the non-obvious bits

Workflows will still need approval, but it should be back to all the same defaults as stock nanodbc. Might be interesting to also add boost to the ci matrix.

File: .gitignore [hunk 3]
```diff
@@ -33,6 +33,7 @@ package-lock.json
 *.code-workspace
 
 # Miscelenaous
+.cache
```
Comment (nimish) [inline .gitignore:36]: for `clangd` analysis, it creates a `.cache` folder

**Review by mloskot — APPROVED** [review]

**Comment by mloskot** [comment]
> Workflows will still need approval, but it should be back to all the same defaults as stock nanodbc. 

There is one failure for MySQL test due to SSL issue which I'm not clear about, but I think this can be ignored for this PR.

> Might be interesting to also add boost to the ci matrix.

Yes, indeed.

We switched to GHA from Travis CI and AppVeyor quite recently, so the CI coverage is still a work in progress (see #265).

> Also, the fallbacks for ancient __GNUC__ could be removed given the minimum req of C++14

Good catch! Opened #383

I'm happy to merge this PR. Thank you for your contribution.

File: example/empty.cpp (hunk 2 above)
Comment (nimish) [reply example/empty.cpp]: > Thanks @nimish
> If there are better ways to achieve similar, then please feel free to propose a change.

Honestly, not sure -- this does seem like a decent solution. It's documented in a comment now too

**Comment by mloskot** [comment]
> Do we need the appveyor env var check?

No, we no longer build on AV, last time was two months ago https://ci.appveyor.com/project/nanodbc/nanodbc
Another good catch to clean up.

**Comment by nimish** [comment]
Cool, there's some minor typos to clean up as well. https://dev.mysql.com/doc/refman/8.0/en/connection-options.html#option_general_ssl-mode says `ssl-mode=PREFERRED` is the default and falls back to unencrypted, but this run just breaks anyway

**Pull request merged**